File: lib/Command/Respond.php
<?php

declare(strict_types=1);

// SPDX-FileCopyrightText: 2026 Cristian Casapu
// SPDX-License-Identifier: AGPL-3.0-or-later

namespace OCA\Sentinel\Command;

use OCA\Sentinel\Service\Inventory;
use OCA\Sentinel\Service\Journal;
use OCP\IUserManager;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Output\OutputInterface;

/**
 * The one thing the daemon outside cannot do for itself.
 *
 * When the writing is coming through php-fpm — which on a Nextcloud server is
 * almost always — there is no process worth stopping. php-fpm is what writes
 * everybody's files all day; killing it stops the server for every account on
 * the machine, which would be a far more reliable way to lose a day than most
 * of what is being watched for.
 *
 * What needs to stop is the *session* doing it: the sync client on the laptop
 * that has caught something. That is Nextcloud's to end, and this is how the
 * daemon asks. Through occ rather than the API, so that nothing has to keep a
 * credential on disk, and so that every invocation is visible in the process
 * list.
 */
class Respond extends Command {
	public function __construct(
		private Inventory $inventory,
		private IUserManager $users,
		private Journal $journal,
	) {
		parent::__construct();
	}

	protected function configure(): void {
		$this->setName('sentinel:respond')
			->setDescription('End an account\'s sessions, and optionally disable it')
			->addOption('uid', null, InputOption::VALUE_REQUIRED, 'The account')
			->addOption('disable', null, InputOption::VALUE_NONE, 'Also disable the account')
			->addOption('reason', null, InputOption::VALUE_REQUIRED, 'What prompted this', 'the process watcher');
	}

	protected function execute(InputInterface $input, OutputInterface $output): int {
		$uid = (string)$input->getOption('uid');
		if ($uid === '') {
			$output->writeln('<error>--uid is required</error>');
			return 1;
		}

		$user = $this->users->get($uid);
		if ($user === null) {
			$output->writeln('<error>No such account: ' . $uid . '</error>');
			return 1;
		}

		$reason = (string)$input->getOption('reason');
		$disable = (bool)$input->getOption('disable');

		$ended = $this->inventory->revokeAllTokens($uid);
		$disabled = false;
		if ($disable) {
			$user->setEnabled(false);
			$disabled = true;
		}

		$this->journal->record(
			'sentinel_responded',
			Journal::ALARM,
			$disabled
				? $uid . ' was signed out of everything and disabled, at the request of ' . $reason . '.'
				: $uid . ' was signed out of everything, at the request of ' . $reason . '.',
			subject: $uid,
			actor: 'sentinel-edr',
			address: null,
			detail: ['uid' => $uid, 'sessionsEnded' => $ended, 'disabled' => $disabled, 'reason' => $reason],
			quiet: 0,
		);

		$output->writeln(($ended ? 'sessions ended' : 'sessions could not be ended')
			. ($disabled ? ', account disabled' : ''));
		return $ended ? 0 : 1;
	}
}

File: edr/sentinel_edr/respond.py
# SPDX-FileCopyrightText: 2026 Cristian Casapu
# SPDX-License-Identifier: AGPL-3.0-or-later
"""Stopping it, carefully.

Two rules, and the second one is the important one.

First: a process that has no business writing into the data directory gets
suspended, not killed. Suspending is reversible — the process is frozen exactly
where it stood, nothing it holds is lost, and a person can look at it and then
either kill it or let it go. Killing an encryptor mid-file and killing a backup
script mid-file look the same from here, and only one of those is a mistake you
can undo.

Second: php-fpm is never touched. On a Nextcloud server php-fpm is what writes
the files, all day, for everybody. Stopping it stops Nextcloud for every account
on the machine — which is a far more reliable way to take a server down than
most of what this daemon is watching for. When the writer is php-fpm, the answer
is not on this side of the wall at all: it is to end the *session* doing it, and
that is Nextcloud's to do.
"""

from __future__ import annotations

import os
import signal
import subprocess
from dataclasses import dataclass

SUSPEND = "suspend"
KILL = "kill"
REPORT = "report"


@dataclass(frozen=True)
class Action:
    what: str
    pids: list[int]
    detail: str = ""


def stop(pids: list[int], how: str, protected: set[int]) -> Action:
    """Suspend or kill the processes that are not allowed to be doing this."""
    if how == REPORT or not pids:
        return Action(what=REPORT, pids=[], detail="reported only")

    touched: list[int] = []
    for pid in pids:
        if pid in protected or pid <= 1 or pid == os.getpid():
            continue
        try:
            os.kill(pid, signal.SIGSTOP if how == SUSPEND else signal.SIGKILL)
            touched.append(pid)
        except (ProcessLookupError, PermissionError):
            continue

    if not touched:
        return Action(what=REPORT, pids=[], detail="nothing left to stop")
    return Action(what=how, pids=touched)


def end_sessions(occ: list[str], as_user: str, uid: str, lock: bool, timeout: int = 60) -> str:
    """Ask Nextcloud to end that account's sessions, and optionally disable it.

    Through occ rather than the API, so that no credential has to be stored
    anywhere on this machine and every invocation is visible in the process
    list.

    occ refuses to run as root — it would leave files in the data directory that
    the web server cannot then touch — so the daemon drops to the account that
    owns the installation before running it. Dropping privilege, never taking
    it: this is deliberately the opposite of asking sudo, which cannot work
    under NoNewPrivileges and would need a sudoers rule that is itself a way in.
    """
    if not occ:
        return "not configured"

    command = [*occ, "sentinel:respond", "--uid", uid]
    if lock:
        command.append("--disable")

    extra: dict = {}
    if as_user:
        try:
            import pwd

            account = pwd.getpwnam(as_user)
        except KeyError:
            return f"failed: no account called {as_user} on this machine"
        extra = {
            "user": account.pw_uid,
            "group": account.pw_gid,
            "env": {
                "HOME": account.pw_dir,
                "USER": as_user,
                "PATH": "/usr/local/bin:/usr/bin:/bin",
            },
        }

    try:
        finished = subprocess.run(
            command, capture_output=True, text=True, timeout=timeout, check=False, **extra,
        )
    except (OSError, subprocess.TimeoutExpired) as error:
        return f"failed: {error}"
    if finished.returncode != 0:
        return f"failed: {(finished.stderr or finished.stdout).strip()[:200]}"
    return (finished.stdout or "done").strip()[:200]
